File: part2.py

START = 50

class Dial:
    def __init__(self, start):
        self.position = start
        self.max = 99
        self.min = 0
        self.points_at_zero = 0

    def turn(self, direction, clicks):
        if direction == "L":
            for i in range(clicks):
                self.position -= 1
                if self.position == 0:
                    self.points_at_zero += 1
                if self.position < self.min:
                    self.position = self.max
        elif direction == "R":
            for i in range(clicks):
                self.position += 1
                if self.position > self.max:
                    self.position = self.min
                    self.points_at_zero += 1
   
    def get_position(self):
        return self.position


def parse_line(line):
    directon, clicks = line.strip()[0], int(line.strip()[1:])
    return directon, clicks

if __name__ == "__main__":

    dial_points_at_zero = 0

    with open("input.txt", "r") as file:
        lines = file.readlines()
        dial = Dial(START)
        for line in lines:
            direction, clicks = parse_line(line)
            dial.turn(direction, clicks)

    print(f"Password is: {dial.points_at_zero}")

File: part2.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"strings"
)

func parseLine(line string) (string, int) {
	direction := string(line[0])
	clicks, err := strconv.Atoi(strings.TrimSpace(string(line[1:])))
	if err != nil {
		fmt.Println("Error parsing clicks:", err)
		return "", 0
	}
	return direction, clicks
}

func main() {

	start := 50
	max := 99
	min := 0
	pointsAtZero := 0

	lines, err := os.ReadFile("input.txt")

	if err != nil {
		fmt.Println("Error reading file:", err)
		return
	}
	for _, line := range strings.Split(string(lines), "\n") {
		direction, clicks := parseLine(line)

		if direction == "L" {
			for i := 0; i < clicks; i++ {
				start -= 1
				if start == 0 {
					pointsAtZero += 1
				}
				if start < min {
					start = max
				}
			}
		} else if direction == "R" {
			for i := 0; i < clicks; i++ {
				start += 1
				if start > max {
					start = min
					pointsAtZero += 1
				}
			}
		}
	}
	fmt.Println("Points at 0: " + strconv.Itoa(pointsAtZero))
}
